RFC-0004-expression-eval.md

```diff
@@ -0,0 +1,80 @@
+
+Unfenced functions may often be implemented to provide an efficient set of functions that are common to a business unit or group.  They allow the users of Presto to efficiently customize Presto into a different flavor specific to their business needs.
+
+A key feature of unfenced functions is that they can be used to implement constant folding.  This is because the function can be evaluated at query planning time, and the result can be used as a constant in the query plan.  This can be a significant performance improvement, as it can reduce the amount of work that needs to be done at query execution time.  For example, if an unfenced function is evaluated over a partition column, the result can be used to prune partitions before the query is executed.
+
+Currently in Presto C++, there is no constant folding support for functions which are written in C++.  In practice, constant folding works for functions which have a corresponding Java implementation, but not for functions which are implemented in C++.  It works by essentially just using a forked set of functions written in Java to evaluate the constant expressions, and then during execution relying on the C++ functions to evaluate the non-constant expressions.  This has a few drawbacks:
+
+1) It requires a forked set of functions to be written in Java, which is a maintenance burden.
+2) An unfenced function written in C++ must have a corresponding Java implementation in order to work properly.
+
+RFC-0003 proposed a mechanism to both add a function registry to allow C++ implemented functions to be exposed to the coordinator without a corresponding Java implementation, and to add constant folding support via remote function evaluation.  This RFC has a similar goal, but represents an improvement over that approach.
+
+In this RFC, we propose to optimize row expressions as a whole, not just function calls.  This has a few advantages:
+
+1) It allows for constant folding to be applied to expressions which are not function calls, such as special form expressions.
+2) It can reduce the overhead of constant folding by allowing the evaluation engine to evaluate the expression as a whole, rather than evaluating each function call individually.
+
+### [Optional] Goals
+
+* Allow for a pluggable implementation of expression evaluation and use this in the Presto optimizer for constant folding
+* Allow connector authors to use this plugin to optimize their own expressions (for example, for filter pushdown)
+
+### [Optional] Non-goals
+
+## Proposed Implementation
+
+A new SPI will be added to Presto to allow for a pluggable implementation of expression evaluation.  A default implementation of this will be added which delegates to the existing expresion evaluation code, but this may be overridden by a custom plugin.
```

Comment:
@mbasmanova @kgpai @kagamiori I wonder if somehow this SPI could also be used to validate fuzzed expressions in Velox using the Presto Java eval. We could probably use the same API but in a different direction - having the C++ code calling a Java backend to evaluate the expression. 

@tdcmeehan could this be also supported in by this API/SPI?

Reply:
@pedroerp Pedro, it is not clear to me what would be the benefit of this over the current approach or using PrestoQueryRunner. Would you clarify?

Reply:
I think it could, but we are actually looking to see if we can use the Velox expression fuzzer to do an A/B test between the Java-only expression optimizer and the Presto C++-backed expression optimizer, since a lot of the work has already been done with this fuzzer and we would prefer not to write it from scratch.

Reply:
@pedroerp My understanding is that PQR already works for any query, but it needs to be integrated into Expression Fuzzer in a similar way it was integrated into Aggregation Fuzzer.